[system] Disable theme recalculation as default behavior (#45405)
mui/material-ui: docs/data/material/migration/upgrade-to-v7/upgrade-to-v7.md

The default `data-testid` prop has been removed from the icons in `@mui/icons-material` in production bundles. This change ensures that the `data-testid` prop is only defined where needed, reducing the potential for naming clashes and removing unnecessary properties in production.

### Theme behavior changes

When CSS theme variables is enabled with built-in light and dark color schemes:

```js
const theme = createTheme({
  cssVariables: {
    colorSchemeSelector: 'class',
  },
  colorSchemes: {
    light: true,
    dark: true,
  },
});
console.log(theme.palette.mode); // 'light'
```

The theme no longer changes between modes:

```js
const { setMode } = useColorScheme();
setMode('dark');

> console.log(theme.palette.mode); // 'light'

Reviewer: How would someone implement a color mode toggler if they can't use `useColorScheme` to get the current mode? 🤔 It's extremely confusing doing `setMode('dark')` and not getting the updated value.

---

Off-topic: I just realized the hook is named `useColorScheme` but it returns `mode` and `setMode`, why is that?

Author: Those are separate thing, the `mode` you get from `useColorScheme` is a React state that ThemeProvider created internally.

When the `mode` state changes, ThemeProvider recalculate the theme as pass to Emotion theme context, causing all components to rerender (this is the previous behavior).

> It's extremely confusing doing setMode('dark') and not getting the updated value.
This is not true, because when you call `setMode('dark')` the `mode` state always update. It's just that the theme now does not change.

May be a blog post or a doc about it might help. I think moving from JS theme to CSS vars theme is quite a shift, I will try to draft the doc about the transition.

Reviewer: If I want to build a light/dark toggle using `useColorScheme` or if I want to have conditional rendering based on the current color scheme (mode), how can I do it if the `mode` returned from `useColorScheme` doesn't return the updated value?

Author: The `mode` returned from `useColorScheme` always updated (it's a React state). It's independent from the theme.

Reviewer: I think there's a typo then? I was confused by the code snippet.

```suggestion
const { setMode } = useColorScheme();
setMode('dark');
console.log(theme.palette.mode); // 'dark'
```

Reviewer: I mean, the snippet is correct (the logged value will be `'light'` until the component re-renders) but it's confusing, especially in conjunction with the phrase preceding the code snippet ("The theme no longer changes between modes"). Can we change the snippet and the preceding phrase?

Author: I see, it's confusing indeed. I updated the snippet and add more details to it. The key here is to show that `mode` from `useColorScheme` is always updated with the setter but the theme object does not change between modes.

```

This default behavior was made to improve performance by avoiding unnecessary re-renders when the mode changes.

It's recommended to use the `theme.vars.*` as values in your styles to refer to the CSS variables directly:

```js
const Custom = styled('div')(({ theme }) => ({
  color: theme.vars.palette.text.primary,
  background: theme.vars.palette.primary.main,
}));
```

> If you need to do runtime calculation, we recommend using CSS approach whenever possible.

Reviewer: ```suggestion
If you need to do runtime calculations, we recommend using CSS instead of JS whenever possible.
```

For example, adjusting the alpha channel of a color can be done using the [`color-mix` function](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/color-mix):

```js
const Custom = styled('div')(({ theme }) => ({
  color: `color-mix(in srgb, ${theme.vars.palette.text.primary}, transparent 50%)`,
}));
```

However, if CSS approach is not possible, you can access the value directly from the `theme.colorSchemes` object, then apply both light and dark styles:

```js
const Custom = styled('div')(({ theme }) => ({
  color: alpha(theme.colorSchemes.light.palette.text.primary, 0.5),
  ...theme.applyStyles('dark', {
    color: alpha(theme.colorSchemes.dark.palette.text.primary, 0.5),
  }),
}));
```

> If any of the methods above do not suit your project, you can opt out from this behavior by specifying `forceThemeRerender` prop to the ThemeProvider component:

Reviewer: ```suggestion
If any of the methods above do not suit your project, you can opt out from this behavior by passing the `forceThemeRerender` prop to the ThemeProvider component:
```
